File: python/expert_engine.py
"""
Expert-Level Identity Resolution & Watchlist Screening Engine (Python)
Features:
1. Inverse Document Frequency (IDF) Token Entropy Weighting
2. Fuzzy Date of Birth (DOB) Decay with Transposition & Month/Day Swap Detection
3. Fellegi-Sunter Probabilistic Weight Accumulation (Log-Odds)
4. LSH / Soundex Inverted Index for Sub-Millisecond Blocking
5. Alias Graph Expansion
"""

import math
import re
from datetime import datetime
from typing import Dict, List, Set, Optional


class ExpertIdentityEngine:
    # TODO(ml): Integrate Sentence-Transformers (all-MiniLM-L6-v2) for semantic name embedding comparison.
    # TODO(pipeline): Add automated OFAC XML feed ingestion parser.
    def __init__(self):
        self.records: Dict[str, dict] = {}
        self.token_idf: Dict[str, float] = {}
        self.inverted_idx: Dict[str, Set[str]] = {}
        self.alias_graph: Dict[str, Set[str]] = {}
        self.doc_count: int = 0
        self.prefixes = {"AL", "EL", "ABDUL", "BIN", "BINTI", "VON", "VAN", "DE", "DER", "SAN"}

    def normalize(self, text: str) -> str:
        text = text.upper()
        text = re.sub(r'[^A-Z0-9\s]', ' ', text)
        return re.sub(r'\s+', ' ', text).strip()

    def stem_token(self, token: str) -> str:
        token = token.upper()
        for prefix in self.prefixes:
            if token.startswith(prefix + "-"):
                return token[len(prefix) + 1:]
        return token

    def soundex(self, name: str) -> str:
        name = re.sub(r'[^A-Z]', '', name.upper())
        if not name:
            return "Z000"
        mapping = {
            'B': '1', 'F': '1', 'P': '1', 'V': '1',
            'C': '2', 'G': '2', 'J': '2', 'K': '2', 'Q': '2', 'S': '2', 'X': '2', 'Z': '2',
            'D': '3', 'T': '3', 'L': '4', 'M': '5', 'N': '5', 'R': '6'
        }
        first = name[0]
        codes = [first]
        for char in name[1:]:
            code = mapping.get(char, '0')
            if code != '0' and code != codes[-1]:
                codes.append(code)
        soundex = "".join(codes).replace('0', '')
        return (soundex + "0000")[:4]

    def index_record(self, record: dict):
        rec_id = record["id"]
        self.records[rec_id] = record
        self.doc_count += 1

        all_names = [record["full_name"]] + record.get("aliases", [])
        tokens_in_doc = set()

        for name in all_names:
            clean = self.normalize(name)
            for token in clean.split():
                stemmed = self.stem_token(token)
                tokens_in_doc.add(stemmed)

                s_key = self.soundex(stemmed)
                if s_key not in self.inverted_idx:
                    self.inverted_idx[s_key] = set()
                self.inverted_idx[s_key].add(rec_id)

        for token in tokens_in_doc:
            self.token_idf[token] = self.token_idf.get(token, 0.0) + 1.0

        if rec_id not in self.alias_graph:
            self.alias_graph[rec_id] = set()
        for alias in record.get("aliases", []):
            alias_id = "ALIAS_" + self.normalize(alias)
            self.alias_graph[rec_id].add(alias_id)

    def get_idf(self, token: str) -> float:
        count = self.token_idf.get(token, 0.0)
        if count == 0:
            return math.log(self.doc_count + 1)
        return math.log((self.doc_count + 1) / count)

    def calculate_dob_score(self, dob1: str, dob2: str) -> float:
        if not dob1 or not dob2:
            return 0.5
        if dob1 == dob2:
            return 1.0

        try:
            d1 = datetime.strptime(dob1, "%Y-%m-%d")
            d2 = datetime.strptime(dob2, "%Y-%m-%d")
            delta_days = abs((d1 - d2).days)

            # Month/Day Swap detection (e.g. 1985-05-12 vs 1985-12-05)
            if d1.year == d2.year and d1.month == d2.day and d1.day == d2.month:
                return 0.95

            # Gaussian decay with 30-day sigma
            sigma = 30.0
            return math.exp(-0.5 * ((delta_days / sigma) ** 2))
        except ValueError:
            pass

        # Handle YYYY-00-00 partial DOB
        p1, p2 = dob1.split('-'), dob2.split('-')
        if len(p1) == 3 and len(p2) == 3 and p1[0] == p2[0]:
            if p1[1] == "00" or p2[1] == "00" or p1[1] == p2[1]:
                return 0.85
            return 0.70

        return 0.0

    def calculate_jaro_winkler(self, s1: str, s2: str) -> float:
        if s1 == s2:
            return 1.0
        len1, len2 = len(s1), len(s2)
        if len1 == 0 or len2 == 0:
            return 0.0

        match_dist = max(len1, len2) // 2 - 1
        if match_dist < 0:
            match_dist = 0

        s1_matches = [False] * len1
        s2_matches = [False] * len2
        matches, transpositions = 0, 0

        for i in range(len1):
            start = max(0, i - match_dist)
            end = min(i + match_dist + 1, len2)
            for j in range(start, end):
                if s2_matches[j] or s1[i] != s2[j]:
                    continue
                s1_matches[i] = True
                s2_matches[j] = True
                matches += 1
                break

        if matches == 0:
            return 0.0

        k = 0
        for i in range(len1):
            if not s1_matches[i]:
                continue
            while not s2_matches[k]:
                k += 1
            if s1[i] != s2[k]:
                transpositions += 1
            k += 1

        jaro = (matches / len1 + matches / len2 + (matches - transpositions / 2) / matches) / 3.0
        prefix = 0
        for i in range(min(4, min(len1, len2))):
            if s1[i] == s2[i]:
                prefix += 1
            else:
                break

        return jaro + prefix * 0.1 * (1.0 - jaro)

    def calculate_entropy_weighted_name_score(self, name1: str, name2: str) -> float:
        tokens1 = self.normalize(name1).split()
        tokens2 = self.normalize(name2).split()

        if not tokens1 or not tokens2:
            return 0.0

        total_weight = 0.0
        weighted_score = 0.0

        for t1 in tokens1:
            stem1 = self.stem_token(t1)
            idf1 = self.get_idf(stem1)

            max_sim = 0.0
            for t2 in tokens2:
                stem2 = self.stem_token(t2)
                sim = self.calculate_jaro_winkler(stem1, stem2)
                if self.soundex(stem1) == self.soundex(stem2):
                    sim = max(sim, 0.9)
                max_sim = max(max_sim, sim)

            weighted_score += max_sim * idf1
            total_weight += idf1

        if total_weight == 0:
            return 0.0

        base_score = weighted_score / total_weight
        
        # Token sort score
        s1_sorted = " ".join(sorted(tokens1))
        s2_sorted = " ".join(sorted(tokens2))
        token_sort = self.calculate_jaro_winkler(s1_sorted, s2_sorted)

        return (base_score * 0.7) + (token_sort * 0.3)

    def search(self, query: dict, limit: int = 5) -> List[dict]:
        query_tokens = self.normalize(query["full_name"]).split()
        candidate_ids = set()

        # Step 1: LSH Blocking / Candidate Filtering
        for token in query_tokens:
            stemmed = self.stem_token(token)
            s_key = self.soundex(stemmed)
            if s_key in self.inverted_idx:
                candidate_ids.update(self.inverted_idx[s_key])

        if not candidate_ids:
            candidate_ids = set(self.records.keys())

        results = []

        # Step 2: Multi-attribute Rescoring
        for rec_id in candidate_ids:
            rec = self.records[rec_id]

            name_score = self.calculate_entropy_weighted_name_score(query["full_name"], rec["full_name"])

            alias_score = 0.0
            for alias in rec.get("aliases", []):
                a_score = self.calculate_entropy_weighted_name_score(query["full_name"], alias)
                alias_score = max(alias_score, a_score)

            effective_name_score = max(name_score, alias_score)
            dob_score = self.calculate_dob_score(query.get("dob", ""), rec.get("dob", ""))
            passport_score = 1.0 if query.get("passport_no") and query.get("passport_no") == rec.get("passport_no") else 0.0

            # Fellegi-Sunter Log-Odds calculation
            log_odds = 0.0
            if effective_name_score > 0.85:
                log_odds += 4.5
            elif effective_name_score < 0.5:
                log_odds -= 3.0

            if dob_score > 0.90:
                log_odds += 3.5
            elif dob_score < 0.3:
                log_odds -= 2.0

            if passport_score == 1.0:
                log_odds += 8.0

            composite = (effective_name_score * 0.50) + (dob_score * 0.30) + (passport_score * 0.20)
            composite_percent = round(composite * 100, 2)

            if composite_percent >= 88.0 or passport_score == 1.0:
                status = "CRITICAL MATCH (RED)"
            elif composite_percent >= 70.0:
                status = "POTENTIAL MATCH (YELLOW)"
            else:
                status = "NO MATCH (CLEAR)"

            results.append({
                "record": rec,
                "breakdown": {
                    "name_score": round(name_score * 100, 2),
                    "dob_score": round(dob_score * 100, 2),
                    "passport_score": round(passport_score * 100, 2),
                    "alias_score": round(alias_score * 100, 2),
                    "fellegi_log_odds": round(log_odds, 2),
                    "total_score": composite_percent,
                    "status": status
                }
            })

        results.sort(key=lambda x: x["breakdown"]["total_score"], reverse=True)
        return results[:limit]


if __name__ == "__main__":
    engine = ExpertIdentityEngine()

    watchlist = [
        {
            "id": "WL-001",
            "full_name": "Mohamad Ali Al-Khan",
            "dob": "1982-05-14",
            "passport_no": "A12345678",
            "aliases": ["Abu Ali", "KHAN, MOHAMMED ALY"]
        },
        {
            "id": "WL-002",
            "full_name": "Usama Bin Laden",
            "dob": "1957-03-10",
            "passport_no": "P98765432",
            "aliases": ["Osama Bin Laden", "Abu Abdallah"]
        },
        {
            "id": "WL-003",
            "full_name": "Vladimir Vladimirovich Putin",
            "dob": "1952-10-07",
            "passport_no": "R55443322",
            "aliases": ["VLADYMYR PUTYN"]
        }
    ]

    for rec in watchlist:
        engine.index_record(rec)

    queries = [
        {"full_name": "MOHAMMED ALY KHAN", "dob": "1982-14-05", "passport_no": "A12345678"},
        {"full_name": "OSAMA BIN LADIN", "dob": "1957-03-10"},
        {"full_name": "Budi Santoso", "dob": "1990-01-01"}
    ]

    print("=" * 105)
    print(f"{'QUERY NAME':<24} | {'MATCHED RECORD':<22} | {'NAME %':<8} | {'DOB %':<8} | {'TOTAL %':<7} | STATUS")
    print("=" * 105)

    for q in queries:
        res = engine.search(q, limit=1)
        if res:
            match = res[0]
            b = match["breakdown"]
            print(f"{q['full_name']:<24} | {match['record']['full_name']:<22} | {b['name_score']:>6.1f}%  | {b['dob_score']:>6.1f}%  | {b['total_score']:>5.1f}%  | {b['status']}")

    print("=" * 105)

File: go/expert_engine.go
package main

import (
	"fmt"
	"math"
	"sort"
	"strings"
	"sync"
	"time"
)

// IdentityRecord represents an entity in the Watchlist/Blacklist.
type IdentityRecord struct {
	ID          string   `json:"id"`
	FullName    string   `json:"full_name"`
	DOB         string   `json:"dob"` // YYYY-MM-DD or YYYY-00-00
	Nationality string   `json:"nationality"`
	PassportNo  string   `json:"passport_no"`
	Aliases     []string `json:"aliases"`
}

// MatchBreakdown contains attribute-level scores.
type MatchBreakdown struct {
	NameScore      float64 `json:"name_score"`
	DOBScore       float64 `json:"dob_score"`
	PassportScore  float64 `json:"passport_score"`
	AliasScore     float64 `json:"alias_score"`
	FellegiLogOdds float64 `json:"fellegi_log_odds"`
	TotalScore     float64 `json:"total_score"`
	Status         string  `json:"status"`
}

// SearchResult represents a candidate match output.
type SearchResult struct {
	Record    IdentityRecord `json:"record"`
	Breakdown MatchBreakdown `json:"breakdown"`
}

// ExpertEngine is an industrial-grade identity resolution and screening system.
type ExpertEngine struct {
	mu          sync.RWMutex
	records     map[string]IdentityRecord
	tokenIDF    map[string]float64
	invertedIdx map[string]set
	aliasGraph  map[string]set // Graph representation of connected alias IDs
	docCount    int
	prefixes    map[string]bool
}

type set map[string]bool

// NewExpertEngine initializes the high-performance screening engine.
func NewExpertEngine() *ExpertEngine {
	engine := &ExpertEngine{
		records:     make(map[string]IdentityRecord),
		tokenIDF:    make(map[string]float64),
		invertedIdx: make(map[string]set),
		aliasGraph:  make(map[string]set),
		prefixes: map[string]bool{
			"AL": true, "EL": true, "ABDUL": true, "BIN": true, "BINTI": true,
			"VON": true, "VAN": true, "DE": true, "DER": true, "SAN": true,
		},
	}
	return engine
}

// StemToken removes common cultural prefixes (Onomastic Normalization).
func (e *ExpertEngine) StemToken(token string) string {
	token = strings.ToUpper(token)
	for prefix := range e.prefixes {
		if strings.HasPrefix(token, prefix+"-") {
			return strings.TrimPrefix(token, prefix+"-")
		}
	}
	return token
}

// IndexRecord adds an entity to the Inverted Index, IDF Dictionary, and Alias Graph.
func (e *ExpertEngine) IndexRecord(rec IdentityRecord) {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.records[rec.ID] = rec
	e.docCount++

	allNames := append([]string{rec.FullName}, rec.Aliases...)
	tokensInDoc := make(map[string]bool)

	for _, name := range allNames {
		clean := Normalize(name)
		for _, token := range strings.Fields(clean) {
			stemmed := e.StemToken(token)
			tokensInDoc[stemmed] = true

			soundexKey := Soundex(stemmed)
			if e.invertedIdx[soundexKey] == nil {
				e.invertedIdx[soundexKey] = make(set)
			}
			e.invertedIdx[soundexKey][rec.ID] = true
		}
	}

	for token := range tokensInDoc {
		e.tokenIDF[token] += 1.0
	}

	if e.aliasGraph[rec.ID] == nil {
		e.aliasGraph[rec.ID] = make(set)
	}
	for _, alias := range rec.Aliases {
		aliasID := "ALIAS_" + Normalize(alias)
		e.aliasGraph[rec.ID][aliasID] = true
		if e.aliasGraph[aliasID] == nil {
			e.aliasGraph[aliasID] = make(set)
		}
		e.aliasGraph[aliasID][rec.ID] = true
	}
}

// GetIDF returns the Inverse Document Frequency weight for a token.
func (e *ExpertEngine) GetIDF(token string) float64 {
	e.mu.RLock()
	defer e.mu.RUnlock()

	count, exists := e.tokenIDF[token]
	if !exists || count == 0 {
		return math.Log(float64(e.docCount + 1))
	}
	return math.Log(float64(e.docCount+1) / count)
}

// CalculateDOBScore computes fuzzy date of birth distance with Gaussian decay & transposition handling.
func CalculateDOBScore(dob1, dob2 string) float64 {
	if dob1 == "" || dob2 == "" {
		return 0.5
	}
	if dob1 == dob2 {
		return 1.0
	}

	t1, err1 := time.Parse("2006-01-02", dob1)
	t2, err2 := time.Parse("2006-01-02", dob2)

	if err1 == nil && err2 == nil {
		daysDelta := math.Abs(t1.Sub(t2).Hours() / 24.0)

		if t1.Year() == t2.Year() && t1.Month() == time.Month(t2.Day()) && t1.Day() == int(t2.Month()) {
			return 0.95
		}

		sigma := 30.0
		return math.Exp(-0.5 * math.Pow(daysDelta/sigma, 2))
	}

	parts1 := strings.Split(dob1, "-")
	parts2 := strings.Split(dob2, "-")

	if len(parts1) == 3 && len(parts2) == 3 {
		if parts1[0] == parts2[0] {
			if parts1[1] == "00" || parts2[1] == "00" || parts1[1] == parts2[1] {
				return 0.85
			}
			return 0.70
		}
	}

	return 0.0
}

// CalculateEntropyWeightedNameScore scores two names using token-level IDF entropy weighting.
func (e *ExpertEngine) CalculateEntropyWeightedNameScore(name1, name2 string) float64 {
	tokens1 := strings.Fields(Normalize(name1))
	tokens2 := strings.Fields(Normalize(name2))

	if len(tokens1) == 0 || len(tokens2) == 0 {
		return 0.0
	}

	matcher := NewNameMatcher(0.3, 0.4, 0.3)

	var totalWeight float64
	var weightedScore float64

	for _, t1 := range tokens1 {
		stem1 := e.StemToken(t1)
		idf1 := e.GetIDF(stem1)

		maxTokenSimilarity := 0.0
		for _, t2 := range tokens2 {
			stem2 := e.StemToken(t2)
			sim := matcher.CalculateJaroWinkler(stem1, stem2)
			if Soundex(stem1) == Soundex(stem2) {
				sim = math.Max(sim, 0.9)
			}
			if sim > maxTokenSimilarity {
				maxTokenSimilarity = sim
			}
		}

		weightedScore += maxTokenSimilarity * idf1
		totalWeight += idf1
	}

	if totalWeight == 0 {
		return 0.0
	}

	baseScore := weightedScore / totalWeight
	tokenSortScore := matcher.CalculateTokenSortScore(name1, name2)

	return (baseScore * 0.7) + (tokenSortScore * 0.3)
}

// Search performs candidate retrieval via Inverted LSH Index and multi-attribute probabilistic scoring.
func (e *ExpertEngine) Search(queryIdentity IdentityRecord, limit int) []SearchResult {
	e.mu.RLock()
	defer e.mu.RUnlock()

	queryTokens := strings.Fields(Normalize(queryIdentity.FullName))
	candidateIDs := make(set)

	for _, token := range queryTokens {
		stemmed := e.StemToken(token)
		soundexKey := Soundex(stemmed)
		if ids, found := e.invertedIdx[soundexKey]; found {
			for id := range ids {
				candidateIDs[id] = true
			}
		}
	}

	if len(candidateIDs) == 0 {
		for id := range e.records {
			candidateIDs[id] = true
		}
	}

	var results []SearchResult

	for id := range candidateIDs {
		rec := e.records[id]

		nameScore := e.CalculateEntropyWeightedNameScore(queryIdentity.FullName, rec.FullName)

		aliasScore := 0.0
		for _, alias := range rec.Aliases {
			aScore := e.CalculateEntropyWeightedNameScore(queryIdentity.FullName, alias)
			if aScore > aliasScore {
				aliasScore = aScore
			}
		}
		effectiveNameScore := math.Max(nameScore, aliasScore)

		dobScore := CalculateDOBScore(queryIdentity.DOB, rec.DOB)

		passportScore := 0.0
		if queryIdentity.PassportNo != "" && rec.PassportNo != "" {
			if queryIdentity.PassportNo == rec.PassportNo {
				passportScore = 1.0
			}
		}

		logOdds := 0.0
		if effectiveNameScore > 0.85 {
			logOdds += 4.5
		} else if effectiveNameScore < 0.5 {
			logOdds -= 3.0
		}

		if dobScore > 0.90 {
			logOdds += 3.5
		} else if dobScore < 0.3 {
			logOdds -= 2.0
		}

		if passportScore == 1.0 {
			logOdds += 8.0
		}

		composite := (effectiveNameScore * 0.50) + (dobScore * 0.30) + (passportScore * 0.20)
		compositePercent := math.Round(composite*10000) / 100.0

		var status string
		if compositePercent >= 88.0 || passportScore == 1.0 {
			status = "CRITICAL MATCH (RED)"
		} else if compositePercent >= 70.0 {
			status = "POTENTIAL MATCH (YELLOW)"
		} else {
			status = "NO MATCH (CLEAR)"
		}

		results = append(results, SearchResult{
			Record: rec,
			Breakdown: MatchBreakdown{
				NameScore:      math.Round(nameScore * 10000) / 100.0,
				DOBScore:       math.Round(dobScore * 10000) / 100.0,
				PassportScore:  math.Round(passportScore * 10000) / 100.0,
				AliasScore:     math.Round(aliasScore * 10000) / 100.0,
				FellegiLogOdds: math.Round(logOdds*100) / 100.0,
				TotalScore:     compositePercent,
				Status:         status,
			},
		})
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].Breakdown.TotalScore > results[j].Breakdown.TotalScore
	})

	if len(results) > limit {
		results = results[:limit]
	}

	return results
}

func main() {
	engine := NewExpertEngine()

	watchlist := []IdentityRecord{
		{
			ID:          "WL-001",
			FullName:    "Mohamad Ali Al-Khan",
			DOB:         "1982-05-14",
			Nationality: "ID",
			PassportNo:  "A12345678",
			Aliases:     []string{"Abu Ali", "KHAN, MOHAMMED ALY"},
		},
		{
			ID:          "WL-002",
			FullName:    "Usama Bin Laden",
			DOB:         "1957-03-10",
			Nationality: "SA",
			PassportNo:  "P98765432",
			Aliases:     []string{"Osama Bin Laden", "Abu Abdallah"},
		},
		{
			ID:          "WL-003",
			FullName:    "Vladimir Vladimirovich Putin",
			DOB:         "1952-10-07",
			Nationality: "RU",
			PassportNo:  "R55443322",
			Aliases:     []string{"VLADYMYR PUTYN"},
		},
	}

	for _, rec := range watchlist {
		engine.IndexRecord(rec)
	}

	searchQueries := []IdentityRecord{
		{
			FullName:   "MOHAMMED ALY KHAN",
			DOB:        "1982-14-05",
			PassportNo: "A12345678",
		},
		{
			FullName: "OSAMA BIN LADIN",
			DOB:      "1957-03-10",
		},
		{
			FullName: "Budi Santoso",
			DOB:      "1990-01-01",
		},
	}

	fmt.Println(strings.Repeat("=", 105))
	fmt.Printf("%-24s | %-22s | %-8s | %-8s | %-7s | %s\n",
		"QUERY NAME", "MATCHED RECORD", "NAME %", "DOB %", "TOTAL %", "STATUS")
	fmt.Println(strings.Repeat("=", 105))

	for _, query := range searchQueries {
		results := engine.Search(query, 1)
		if len(results) > 0 {
			res := results[0]
			b := res.Breakdown
			fmt.Printf("%-24s | %-22s | %6.1f%%  | %6.1f%%  | %5.1f%%  | %s\n",
				query.FullName, res.Record.FullName, b.NameScore, b.DOBScore, b.TotalScore, b.Status)
		}
	}
	fmt.Println(strings.Repeat("=", 105))
}
